swap-feed/src/rate.rs

```diff
@@ -107,29 +107,64 @@ impl crate::traits::LatestRate for FixedRate {
     }
 }
 
-/// Produces [`Rate`]s based on [`PriceUpdate`]s from kraken and a configured
-/// spread.
+/// Produces [`Rate`]s based on [`PriceUpdate`]s from kraken and bitfinex,
+/// and a configured spread.
 #[derive(Debug, Clone)]
-pub struct KrakenRate {
+pub struct ExchangeRate {
     ask_spread: Decimal,
-    price_updates: crate::kraken::PriceUpdates,
+    kraken_price_updates: crate::kraken::PriceUpdates,
+    bitfinex_price_updates: crate::bitfinex::PriceUpdates,
+    kucoin_price_updates: crate::kucoin::PriceUpdates,
 }
 
-impl KrakenRate {
-    pub fn new(ask_spread: Decimal, price_updates: crate::kraken::PriceUpdates) -> Self {
+impl ExchangeRate {
+    pub fn new(
+        ask_spread: Decimal,
+        kraken_price_updates: crate::kraken::PriceUpdates,
+        bitfinex_price_updates: crate::bitfinex::PriceUpdates,
+        kucoin_price_updates: crate::kucoin::PriceUpdates,
+    ) -> Self {
         Self {
             ask_spread,
-            price_updates,
+            kraken_price_updates,
+            bitfinex_price_updates,
+            kucoin_price_updates,
         }
     }
 }
 
-impl crate::traits::LatestRate for KrakenRate {
-    type Error = crate::kraken::Error;
+#[derive(Clone, Debug, PartialEq, thiserror::Error)]
+pub enum Error {
+    #[error("Kraken error")]
+    Kraken(#[from] crate::kraken::Error),
+    #[error("Bitfinex error")]
+    Bitfinex(#[from] crate::bitfinex::Error),
+    #[error("KuCoin error")]
+    KuCoin(#[from] crate::kucoin::Error),
+    #[error("Exchanges disagree by more than 10%")]
+    SpreadTooWide,
+}
+
+impl crate::traits::LatestRate for ExchangeRate {
+    type Error = Error;
 
     fn latest_rate(&mut self) -> Result<Rate, Self::Error> {
-        let update = self.price_updates.latest_update()?;
-        let rate = Rate::new(update.ask, self.ask_spread);
+        let kraken_update = self.kraken_price_updates.latest_update()?;
+        let bitfinex_update = self.bitfinex_price_updates.latest_update()?;
+        let kucoin_update = self.kucoin_price_updates.latest_update()?;
+
+        let asks = [kraken_update.ask, bitfinex_update.ask, kucoin_update.ask];
+        let average_ask = asks.into_iter().sum::<bitcoin::Amount>() / (asks.len() as u64);
+        let min_ask = asks.iter().min().expect(">0 asks");
```

Comment:
will this panic if one of the exchanges had zero volume? I'm afraid this might be triggered if one of them pause their orders for maintenance.

Reply:
No, this will only panic if `let asks = [a, b, c]` magically becomes `let asks = []`.